test/example/module-splitting.cpp

```diff
@@ -0,0 +1,208 @@
+#include <cassert>
+#include <iostream>
+
+#include "ir/module-splitting.h"
+#include "ir/stack-utils.h"
+#include "wasm-features.h"
+#include "wasm-printing.h"
+#include "wasm-s-parser.h"
+#include "wasm-validator.h"
+#include "wasm.h"
+
+using namespace wasm;
+
+std::unique_ptr<Module> parse(char* module) {
+  auto wasm = std::make_unique<Module>();
+  wasm->features = FeatureSet::All;
+  try {
+    SExpressionParser parser(module);
+    Element& root = *parser.root;
+    SExpressionWasmBuilder builder(*wasm, *root[0], IRProfile::Normal);
+  } catch (ParseException& p) {
+    p.dump(std::cerr);
+    Fatal() << "error in parsing wasm text";
+  }
+  return wasm;
+}
+
+void do_test(std::set<Name> keptFuncs, std::string module) {
+  WasmValidator validator;
+  bool valid;
+
+  auto primary = parse(&module.front());
+  valid = validator.validate(*primary);
+  assert(valid && "before invalid!");
+
+  std::cout << "Before:\n";
+  WasmPrinter::printModule(primary.get());
+
+  ModuleSplitting::Config config;
+  config.primaryFuncs = keptFuncs;
+  auto secondary = splitFunctions(*primary, config);
+
+  valid = validator.validate(*primary);
+  assert(valid && "after invalid!");
+  valid = validator.validate(*secondary);
+  assert(valid && "secondary invalid!");
+
+  std::cout << "After:\n";
+  WasmPrinter::printModule(primary.get());
+  std::cout << "Secondary:\n";
+  WasmPrinter::printModule(secondary.get());
+  std::cout << "\n\n";
+}
+
+int main() {
+  // Trivial module
+  do_test({}, "(module)");
```

Comment:
Do we not any anything better than hand-rolling for C++ unit tests in binaryen?  i.e. no gtest?

Reply:
Unfortunately not. I wish we had gtest (and LLVM's lit and FileCheck), but overhauling all the testing infrastructure would be a big project. I wouldn't want to do it only half way and leave us with yet another way to run tests. I already accidentally did that when I introduced the Python unit tests.